### PR TITLE
[Laravel] Storage::url() and Storage::temporaryUrl()

### LINKED ISSUE #603: [Laravel/s3] Calling Storage::url() fails
With the storage configured with `'driver' => 'async-aws-s3'`.
Calling this:
```php
$publicUrl = Storage::url($filename);
```

Which is documented here: https://laravel.com/docs/7.x/filesystem#file-urls
Fails with the following error:

```json
{
    "message": "This driver does not support retrieving URLs.",
    "exception": "RuntimeException",
    "file": "/application/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php",
    "line": 440,
    "trace": [
        {
            "file": "/application/app/Services/ProfileImageService.php",
            "line": 18,
            "function": "url",
            "class": "Illuminate\\Filesystem\\FilesystemAdapter",
            "type": "->"
        },
        ...
    ]
}
```

The Flystem-adapter used internally is `AsyncAws\Flysystem\S3\S3FilesystemV1`.

Laravel checks if the adapter has a method `getUrl` or is an instance of some specified adapters and calls special methods for those.
But since `S3FilesystemV1` does not have a `getUrl` or is listed in the special cases it will fail. Or am I missing something?
https://github.com/laravel/framework/blob/7.x/src/Illuminate/Filesystem/FilesystemAdapter.php#L421

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #603

### CODE REVIEW

File: src/Integration/Flysystem/S3/src/S3FilesystemV1.php [hunk 1]
```diff
@@ -84,6 +84,19 @@ public function __construct(S3Client $client, string $bucket, string $prefix = '
...
 
+    public function getClient(): S3Client
+    {
+        return $this->client;
+    }
+
+    /**
+     * Get the S3Client bucket.
+     */
+    public function getBucket(): string
```
Comment (Nyholm) [inline src/Integration/Flysystem/S3/src/S3FilesystemV1.php:95]: The same method exists in Flysystem's S3 adapter.

**Comment by hmazter** [comment]
I will test this and check it out in more depth tomorrow.

But the first thing I'm noticing is that the URL is time-limited which the vanilla Laravel URL is not. `Storage::url()` just returns the plain URL to the object, if the object has a public ACL it is viewable at that URL (for example an image uploaded with public ACL, which is what I use that for)

For signed URLs with expiration `Storage::temporaryUrl()` is used, https://github.com/laravel/framework/blob/61e00dac57ec137c52d6cd1e76de9c374dc6fae6/src/Illuminate/Filesystem/FilesystemAdapter.php#L543

**Comment by Nyholm** [comment]
Oh. Yeah. I’ve implemented the temporary url. :/

**Comment by Nyholm** [comment]
Blocked by https://github.com/async-aws/aws/pull/605

**Comment by Nyholm** [comment]
I am happy with this PR now.

File: src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     *
+     * @throws \RuntimeException
+     */
+    public function url($path)
+    {
+        $url = $this->temporaryUrl($path, new \DateTimeImmutable('+10minutes'));
+
+        // remove all query parameters.
+        if (false === $pos = strpos($url, '?')) {
+            throw new \RuntimeException('Expected presigned URL to include a quiery string');
```
Comment (jderusse) [inline src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php:63]: query

File: src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    private function getFlysystemAdapter(): S3FilesystemV1
+    {
+        if (!method_exists($this->driver, 'getAdapter')) {
+            throw new \RuntimeException(sprintf('Could not call "getAdapter" of class "%s"', \get_class($this->driver)));
+        }
+
+        $adapter = $this->driver->getAdapter();
+
+        if ($adapter instanceof CachedAdapter) {
+            $adapter = $adapter->getAdapter();
```
Comment (jderusse) [inline src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php:78]: Is it possible to have several layers of cache? ie. `MemoryCache(FileCache(ConcreteAdapter))`

Comment (Nyholm) [reply src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php:78]: Yes, it is. Good catch.. I copied the code from Laravel. I didnt think much

**Comment by hmazter** [comment]
It tested this PR and it seems to work. 👍 
Noticed one thing though. The URL returned is a different format then the one returned "natively" by Laravel.

Url returned with this PR:
`https://s3.eu-north-1.amazonaws.com/{bucket-name}/{path}`

Url returned with native Laravel Flysystem:
`https://bucket-name.s3.eu-north-1.amazonaws.com/{path}`

Object URL in S3 Console:
`https://bucket-name.s3.eu-north-1.amazonaws.com/{path}`

It works, just didn't have the new format in our CSP image-src 😄

**Comment by Nyholm** [comment]
The one with the bucket name is "virtual hosted style". I knew one was "old" and one was "new". 

Looking at the docs now: https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAPI.html

It seams like buckets created after 2020-09-30 will only support the virutal host style. I'll better write a ticket so we update. 

Thank you for pointing this out

**Comment by Nyholm** [comment]
https://github.com/async-aws/aws/issues/612

**Pull request merged**